### PR TITLE
 Fixed the issue #811 and #812, which leading the pipeline break

### PR DESCRIPTION
**Pull request by Dedaniya08**
## PR checklist

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/ampliseq/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, also make a PR on the nf-core/ampliseq _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [x] Make sure your code lints (`nf-core pipelines lint`).
- [x] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [x] Check for unexpected warnings in debug mode (`nextflow run . -profile debug,test,docker --outdir <OUTDIR>`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

**Comment by Dedaniya08** [comment]
The issue of pipeline breaking is caused at two steps for issue #811 and #812 . I have added the retry max 3 attempt which has lead to resolving the both the error in both the files
second_step


Both the step are resolved.

**Comment by d4straub** [comment]
Hi there, thanks for the idea. However, this is a very bad idea to retry for any error code. This can lead to a lot of confusion. I oppose that change.
We typically only retry on specific error codes, see https://github.com/nf-core/ampliseq/blob/8f139cecd95a23920617f2c2b154733494ed4562/conf/base.config#L18.

If that change solves you problem thats great, but there is no need to change pipeline code instead you can achieve this also by using `-c azure.conf` that contains:
```
process {
    errorStrategy = 'retry'
}
```
or use more process targeted approach as in e.g. https://github.com/nf-core/ampliseq/blob/8f139cecd95a23920617f2c2b154733494ed4562/conf/base.config#L65-L69

**Comment by Dedaniya08** [comment]
The code is encountering frequent issues. In this PR, step #811 focuses on renaming the files, and step #812 maps the filtered files. I verified the files after a retry, and the output remains consistent. There are no discrepancies in the files before and after applying this PR.

The recurring issue is due to Azure's local environment and the process of extracting files from the data lake. This behavior will persist without the changes introduced here. Therefore, I strongly recommend including this PR to address the problem effectively.

**Comment by d4straub** [comment]
Thanks for explaining your reasoning. I do believe that the resulting files are fine and I acknowledge that this proposed change will solve the issue on Azure. But it will negatively affect the execution of the pipeline on any other platform (retry for all exit codes of `RENAME_RAW_DATA_FILES` & `DADA2_FILTNTRIM`, which is not desirable). I would prefer a fix that doesn't have a negative impact at all.
I'll try to lure some more people here to get more opinions.

File: modules/local/dada2_filtntrim.nf [hunk 1]
```diff
@@ -1,6 +1,8 @@
 process DADA2_FILTNTRIM {
     tag "$meta.id"
     label 'process_medium'
+    errorStrategy 'retry'
+    maxRetries 3
```
Comment (ewels) [inline modules/local/dada2_filtntrim.nf:5]: This is [already set](https://github.com/nf-core/ampliseq/blob/8f139cecd95a23920617f2c2b154733494ed4562/conf/base.config#L19) globally at `process` level, so no need to set again.
```suggestion
```

File: modules/local/rename_raw_data_files.nf [hunk 2]
```diff
@@ -1,6 +1,8 @@
 process RENAME_RAW_DATA_FILES {
     tag "$meta.id"
     label 'process_low'
+    errorStrategy 'retry'
+    maxRetries 3
```
Comment (ewels) [inline modules/local/rename_raw_data_files.nf:5]:
```suggestion
```

**Review by ewels — CHANGES_REQUESTED** [review]
This issue seems related to the Azure infrastructure setup and is not general to all users of the pipeline, or really pipeline specific. I do not think that we should apply blanket `retry` for all users like this.

Alternatives would be setting this on user configs instead, or better still adding to the [`azurebatch` institutional config](https://nf-co.re/configs/azurebatch/) as a [pipeline-specific config](https://github.com/nf-core/configs?tab=readme-ov-file#adding-a-new-pipeline-specific-config).

The latter is already set up in this pipeline so no changes needed here, purely nf-core/configs repo. Need to create a new config file there and then include it [here](https://github.com/nf-core/configs/blob/master/pipeline/ampliseq.config).

**Comment by ewels** [comment]
It's also possible that the `retry` solves the issue here because the retry attempts give the task more memory (the `task.attempt` multiplier). This could be needed when pulling data from a data lake, which is a slightly non-standard method of data staging.

https://github.com/nf-core/ampliseq/blob/8f139cecd95a23920617f2c2b154733494ed4562/conf/base.config#L30

So something else that you could try in the azure-specific pipeline config is to assign more memory to these tasks. Then they might not fail in the first place and you may not need the `retry`.

**Comment by ewels** [comment]
@d4straub - suggest we close this PR and @Dedaniya08 you can open one to nf-core/configs with the relevant changes.

**Comment by d4straub** [comment]
Yes I agree @ewels , better solved in nf-core/configs. Thanks a lot!

**Pull request closed without merging**